keps/sig-node/1999-checkpoint-restore/README.md

```diff
@@ -0,0 +1,277 @@
+
+   * [Task Migration at Google Using CRIU](https://www.linuxplumbersconf.org/event/2/contributions/209/)
+   * [Update on Task Migration at Google Using CRIU](https://linuxplumbersconf.org/event/4/contributions/508/)
+
+   See especially the second presentation for limitations Google is
+   experiencing using CRIU based container migration in production.
+
+ * Podman supports container migration in its simplest form (checkpoint,
+   transfer, restore): https://criu.org/Podman
+
+   The reason for mentioning Podman in this list is that adding
+   checkpoint and restore to Podman can be seen as a preparation for
+   this KEP. Especially as Podman uses CNI for network configuration
+   it required changes to CRIU and runc to allow restoring into a
+   previously set up network namespace instead of letting CRIU handle
+   the network namespace restore:
+
+   * https://github.com/opencontainers/runc/pull/1849
+
+   This work lead to CRIU's support to allow restoring processes
+   out of and into PID namespaces:
+
+   * https://github.com/checkpoint-restore/criu/pull/1056
+   * https://github.com/opencontainers/runc/pull/2525
+
+   As well as the runc support to allow restoring into all kinds of
+   namespaces: https://github.com/opencontainers/runc/pull/2583
+
+   The support to allow restoring into all kinds of namespaces is
+   especially important when checkpointing and restoring containers
+   out of and into pods.
+
+The motivation is to get closer to container migration by taking the first
+step and providing support for simple checkpoint and restore. The motivation
+is existing checkpoint/restore/migration support in other container
+environments showing that it is useful and production ready.
+
+### Goals
+
+The goal of this KEP is to introduce *checkpoint* and *restore* to the API.
+
+### Non-Goals
+
+Out of scope of this KEP are high level discussions about how to implement
+container migration. This is only about low level primitives to add
+*checkpoint* and *restore* to the API.
+
+
+## Proposal
+
+```patch
+--- a/api.proto
++++ b/api.proto
+@@ -87,6 +87,11 @@ service RuntimeService {
+     // Once it returns error, new container log file MUST NOT be created.
+     rpc ReopenContainerLog(ReopenContainerLogRequest) returns (ReopenContainerLogResponse) {}
+ 
++    // CheckpointContainer checkpoints a container
++    rpc CheckpointContainer(CheckpointContainerRequest) returns (CheckpointContainerResponse) {}
++    // RestoreContainer restores a container
++    rpc RestoreContainer(RestoreContainerRequest) returns (RestoreContainerResponse) {}
++
+     // ExecSync runs a command in a container synchronously.
+     rpc ExecSync(ExecSyncRequest) returns (ExecSyncResponse) {}
+     // Exec prepares a streaming endpoint to execute a command in the container.
+@@ -1308,3 +1313,23 @@ message ReopenContainerLogRequest {
+ 
+ message ReopenContainerLogResponse{
+ }
++
++message CheckpointContainerRequest {
```

Comment:
In my opinion it should be possible to specify the location of the checkpoint (also when restoring a checkpoint of course). This would make the protocol much more flexible as it allows to
- create multiple checkpoints
- process the checkpoint after it's been created. This would be especially important for migration/cloning of Pods.

If the actions include a checkpoint path, it might be worth to plan for further extensions in the future. Especially live-migration could require a lot more options if "actual" live-migration should be possible (I don't want to go into details here, as it's not the scope of this KEP).
My suggestion would be to use a struct to pass the parameters, similar to CreateContainer and others. That way it would be easy to extend the protocol in the future without requiring every implementation to adapt immediately, and the interface of cri-api would be kept clean as well.

Reply:
Thanks for having a look and for your feedback. Using a struct sounds like a good idea. Before changing the proposal I will wait, however, if there are further comments.